### DISCUSSION TITLE
Confusion between User Access Permissions & Document Security Restrictions

### DISCUSSION THREAD

**Discussion by eabase** [py-pdf/pypdf#2582] [Q&A]
There seem to be some confusion between `User Access Permissions` and `Document [Security] Restrictions`, when handling permissions. (Or at least I am confused, because it looks like you are mixing up the two.)

- `User Access Permissions` is used for setting permissions on encrypted files requiring a password.
- `Document [Security] Restrictions` can be set on any PDF file (without encryption)

- There is also an ambiguous `DeprecationWarning` that tell you to use another thing, and then doesn't work. 

Looking at some PDF document you can have the `Document [Security] Restrictions` as shown in Adobe Acrobat, **without** any Encryption. 

Here's how my PDF looks. 
(Sorry, I cannot share it as it belongs to a company.)

![Document_Properties2](https://github.com/py-pdf/pypdf/assets/52289379/33e417fe-3c97-46a6-9c38-c5a51a7b92c2)

---

So if I have a **non-encrypted** PDF with the following [restrictions](https://pypdf.readthedocs.io/en/latest/modules/constants.html):  
```bash
EXTRACT_TEXT_AND_GRAPHICS 
ASSEMBLE_DOC
```
(512 and 1024, respectively)

The question is:
  1. *How do I read these from the PDF?*
  2. *How do I change them in the same PDF?*

---

## Environment
Which environment were you using when you encountered the problem?

```bash
# python -m platform
Windows-10-10.0.22000-SP0

# python -c "import pypdf;print(pypdf._debug_versions)"
pypdf==4.1.0, crypt_provider=('cryptography', '42.0.5'), PIL=10.3.0
```

## Code + PDF

This is a REPL example that shows the issue:

```python
import pypdf
from pypdf import PdfReader, PdfWriter
from pypdf.constants import UserAccessPermissions

IFILE = 'some.pdf'
reader = PdfReader(IFILE)

reader.decode_permissions(UserAccessPermissions.all())
# ...
# <stdin>:1: DeprecationWarning: decode_permissions is deprecated and will be removed in pypdf 5.0.0. Use user_access_permissions instead.

reader.user_access_permissions(UserAccessPermissions.all())
# TypeError: 'NoneType' object is not callable


```

**Comment by stefan6419846** [comment]
It seems like you are mixing up something: A PDF file can have a user password and an owner password. The user password is required each time you open the file. An owner password is used to restrict the permissions a regular user (without specifying the owner password) should be allowed to do.

You are correct that `decode_permissions` has been deprecated as this rather is a property on the document. Nevertheless, you are using both functionality wrong (see https://github.com/py-pdf/pypdf/discussions/2389#discussioncomment-8000710 as well): You either want 

* `reader.decode_permissions(reader._encryption.P)` (deprecated) or 
* `reader.user_access_permissions` (flags instance) or `reader.user_access_permissions.to_dict()` (dictionary similar to the one from `decode_permissions()`.

Alternatively, you might pass any set of permission flag value into `UserAccessPermissions()` to get a verbose representation of it.

To change the permissions, use `PdfWriter.encrypt()` with the appropriate parameters.

**Reply by eabase** [reply]
Yes, I am probably mixing up the correct formatting, since the formatting you say, and have documented simply doesn't work:

```python
>>> reader.user_access_permissions.to_dict()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'NoneType' object has no attribute 'to_dict'

>>> reader.decode_permissions(reader._encryption.P)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'NoneType' object has no attribute 'P'
```

Again, to set the `Document Restrictions` (as I showed above), **you don't need any encryption or password**. 
So I don't understand how you got this idea and from where? 
I certainly have nit seen this in any PDF design or standard documentation. 

If you want to prove me wrong, please provide a working code to correctly read [these](https://github.com/py-pdf/pypdf/blob/main/tests/test_reader.py#L724) from **any** PDF:
```python
    base = {
        "accessability": False,  # Do not fix typo, as part of official, but deprecated API.
        "annotations": False,
        "assemble": False,
        "copy": False,
        "forms": False,
        "modify": False,
        "print_high_quality": False,
        "print": False,
    }
```

**Reply by stefan6419846** [reply]
Sorry for the confusion - there is indeed a way to provide restrictions in the catalog dictionary by using the `/Perms` dictionary (table 28). It can be retrieved like this:

```
from pypdf import PdfReader
from pypdf.constants import CatalogDictionary

reader = PdfReader("file.pdf")
perms = reader.root_object.get(CatalogDictionary.PERMS, None)
print(perms)
```

If set, this might have the key `/DocMDP` pointing to a signature reference dictionary (table 258) (or the `/UR3` key, but this is even further related to signatures). Its `P` value represents the permissions:

> If this entry is present, consumer applications shall enforce the permissions specified by the **P** attribute in the **DocMDP** transform parameters dictionary and shall also validate the corresponding signature based on whether any of these permissions have been violated.

This refers to section 12.8.2.2, where we can read that this primarily focuses signatures, which are unsupported by *pypdf* and thus I am not completely sure if there is something we can do about this. Further references point to table 254, where you can read how to interpret these restrictions as well.

**Reply by eabase** [reply]
That doesn't work either ...

```
    perms = reader.root_object.get(CatalogDictionary.PERMS, None)
            ^^^^^^^^^^^^^^^^^^
AttributeError: 'PdfReader' object has no attribute 'root_object'. Did you mean: 'get_object'?
```

I also tried with suggestion: 
```
    perms = reader.get_object(CatalogDictionary.PERMS, None)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: PdfReader.get_object() takes 2 positional arguments but 3 were given
```

Error is completely wrong! 🤣

**Reply by stefan6419846** [reply]
Sorry, I was using the latest unreleased code for researching/testing this. For now, you will have to use

```
from pypdf.constants import TrailerKeys

root = reader.trailer[TrailerKeys.ROOT].get_object()
perms = root.get(CatalogDictionary.PERMS, None)
```

instead to retrieve the `root_object` manually.

**Reply by eabase** [reply]
This returns `None`.

**Reply by stefan6419846** [reply]
Which just means that the key is not set.

Unless you are able to provide a corresponding PDF file, I am going to stop here as this is just tedious guesswork from my side, given the general complexity of the PDF specs. Please either provide a corresponding file for further analysis or analyze the document yourself and document your results.

**Reply by eabase** [reply]
Hi Stefan, 
I understand and thank you for the help and quick response. I don't know how to reproduce the document without revealing private information. It was generated by some Salesforce app, and it has **no fonts** and `Page Extraction` has been disabled. I don't know how to analyze this. This is what security shows from Acrobat Reader. So I am not sure what `key` you refer to?

![Document_Properties3](https://github.com/py-pdf/pypdf/assets/52289379/d3eae52e-dd83-4af5-a9a0-d548449333c2)

Any suggestions how to go about further analyzing this document, so that it would be helpful to you?

**Reply by stefan6419846** [reply]
For your own analysis, consider uncompressing all data streams using `pdftk file.pdf output out.pdf uncompress`. Then you can view the `out.pdf` file inside a regular text editor and look for permission-related aspects, although this might be difficult if encryption comes in.

According to your screenshot, the PDF is indeed encrypted by a password. Nevertheless, for some reason, *pypdf* does not seem to detect this - otherwise `reader._encryption` would not be `None`. If you have a look at `reader.is_encrypted`, you will see that encryption usually is present if `/Encrypt` is not a key inside the trailer dictionary. This is in accordance with the PDF 1.7 specification, section 7.6.1 (or PDF 2.0 specification, section 7.6.2):

> Encryption-related information shall be stored in a document’s *encryption dictionary*, which shall be the value of the **Encrypt** entry in the document’s trailer dictionary [...]. The absence of this entry from the trailer dictionary means that a conforming reader shall consider the document to be not encrypted.

**Reply by eabase** [reply]
Yep, sorry again, I didn't tell you full story. The original files were not encrypted, but I was told (above) that in order to change the permissions it needed to be saved as encrypted. I tried with various document permission masks, from 568,3900, to 4095, as it was unclear what what R13-R31 is doing. So I saved the merged file with:
```python
writer.encrypt(user_password='', owner_password=None, permissions_flag=perms, algorithm="AES-256")
```

...and what you see in the (main) `Document Properties` tab, are the same for both non-encrypted and the encrypted versions. It would be great if you guys could map out the functionality of the masks in those two windows. I really don't understand it. (Which is why I thought I might not be alone and gave this title to this issue.) 

And where are all those other R13-R31 documented? I suspect that `user` and `owner` permissions are kept separate somehow, so may explain the other bits in R13-R31?

What are the "permission-related aspects"?  (I don't see any.)
I tried:
```powershell
pdftk.exe myfile.pdf output unco.pdf uncompress
```

**Reply by stefan6419846** [reply]
> The original files were not encrypted, but I was told (above) that in order to change the permissions it needed to be saved as encrypted.

I am mostly aware of owner passwords which allow restricting the permissions for the user (apart from certificate-based encryption, which *pypdf* does not support). Your original document might use another mechanism, but without doing a full analysis or digging into its details, this is hard to say. Setting a owner password through *pypdf* will most likely have an influence on the results.

> And where are all those other R13-R31 documented? I suspect that `user` and `owner` permissions are kept separate somehow, so may explain the other bits in R13-R31?

The user password is the one required for actually opening the file - you would have stumbled upon this already ;) The owner password is mostly for restricting the permissions as mentioned above.

All permission bits with a name of the pattern `R\d+` are just "reserved" and should be considered with no further meaning. From the PDF 1.7 specification, section 7.6.3.2:

> Conforming readers shall ignore all flags other than those at bit positions 3, 4, 5, 6, 9, 10, 11, and 12.

In this case, bit position 3 maps to `PRINT` (see the table in the specification), which you will find as the value 8 inside `UserAccessPermissions` due to basic binary principles: `8_10 = 100_2 = 2^3`.

> What are the "permission-related aspects"? (I don't see any.)

Usual suspects are dictionary keys like `/P` or `/Perms`, but this surely is no definitive answer.

**Reply by eabase** [reply]
Problem seem to be, knowing the correct mappings. 
Even in Adobe they are using slightly different names for the 2 windows (I described above), while the items in `pypdf` have labels that doesn't correspond to anything. 

![pdf_permissions](https://github.com/py-pdf/pypdf/assets/52289379/7c019a35-22a4-4b94-85de-e7ce665149cb)

**Reply by pubpub-zz** [reply]
@eabase 
these are the permissions as reviewed in PDF 2.0 specification:
![image](https://github.com/py-pdf/pypdf/assets/4083478/966a2158-eb4d-47cb-959a-d84ccacdc21f)

![image](https://github.com/py-pdf/pypdf/assets/4083478/58528ebf-cc31-4653-8889-07dc049126f6)
![image](https://github.com/py-pdf/pypdf/assets/4083478/fd088be3-cb2d-4ae8-8313-c53c8d22fe03)

**Reply by pubpub-zz** [reply]
about "creation of template pages"
https://acrobatusers.com/forum/security/newbie-question-what-does-allow-creation-template-pages-mean/

**Reply by stefan6419846** [reply]
@pubpub-zz The issue in this case seems to be that *pypdf* would not be able to extract them from the initial PDF file, although there seem to indeed be some permissions. The current state is that there is not much we can do about this without the actual PDF file.

**Reply by eabase** [reply]
It seem that since bits R13-32 "Must be 1" we can't use LSB for permissions like saying `perm=3900` or `0b0100`. This complicates the permission mask as the int become very large.

**Reply by stefan6419846** [reply]
You can use every value you like for each bit, but as this violates the PDF spec (which speaks of 32 permission bits and expected states for some bits), there is not much value in it and saving/re-writing the file might auto-magically fix them. Additionally, only your application might have some advantage from it.

**Reply by eabase** [reply]
Sure, that's your decision. At the end of the day I just wanted to enable page extraction (for copy/paste), by setting bits R5,R10, so that it is working as expected in Adobe Acrobat. It still isn't working, and I suspect the tool that exported it to PDF, is exploiting some bug that they use to prevent copy paste, and in addition have disabled usable fonts or maybe by "drawing the text" through streams. There are several streams in that document. 

PS. Saving an encrypted copy of the originally unencrypted copy, and with all enabled permissions, did not work. (Which is why we are having this discussion.)

**Reply by stefan6419846** [reply]
> It still isn't working, and I suspect the tool that exported it to PDF, is exploiting some bug that they use to prevent copy paste, and in addition have disabled usable fonts or maybe by "drawing the text" through streams.

It does not necessarily have to be a bug, but could be a general PDF feature as well. I am just not aware/sure where this restrictions are defined in your case - the PDF specification is quite huge and thus hard to know every detail about it.

Rendering paths into a document instead of using a font-based approach is not too uncommon - even Microsoft Print to PDF tends to do this. This is rather easy to verify by using `page.extract_text()` - this should extract the text if there is any as *pypdf* being a library ignores any restrictions as far as possible. With rendered paths doing OCR is the only viable solution anyway.

**Reply by eabase** [reply]
Indeed, inspecting the raw PDF document, seem to indicate it was created with `Microsoft Print to PDF`. `page.extract_text()` came out empty with no strings at all.

Anyway, thank you for all your help.

**Reply by pubpub-zz** [reply]
> Sure, that's your decision. At the end of the day I just wanted to enable page extraction (for copy/paste), by setting bits R5,R10, so that it is working as expected in Adobe Acrobat. It still isn't working, and I suspect the tool that exported it to PDF, is exploiting some bug that they use to prevent copy paste, and in addition have disabled usable fonts or maybe by "drawing the text" through streams. There are several streams in that document.

The copy and paste from a document may be allowed but can be prevented by many way, at least two described here :
a) instead of using text with fonts rendering, you can replace the text by raw(line and curves, or even pixel prints) printing instructions preventing extractions (eg in #2583)
 => in such cases pypdf extract_text returns no text
b) you can scramble the `/ToUnicode` data preventing or making unreable the text. 
 => in such cases most of the time pypdf extract_text returns unreadable text or spaces

Can't it be on this cases you are facing ?

oups.  already described by @stefan6419846
